CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+<sup>[[link](#no-and-or-or)]</sup>
+
+  ```Ruby
+  # bad
+  # boolean expression
+  if some_condition and some_other_condition
+    do_something
+  end
+
+  # control flow
+  document.saved? or document.save!
+
+  # good
+  # boolean expression
+  if some_condition && some_other_condition
+    do_something
+  end
+
+  # control flow
+  document.saved? || document.save!
+  ```
+
+* <a name="no-multiline-ternary"></a>
+  Avoid multi-line `?:` (the ternary operator); use `if/unless` instead.
+<sup>[[link](#no-multiline-ternary)]</sup>
+
+* <a name="if-as-a-modifier"></a>
+  Favor modifier `if/unless` usage when you have a single-line body. Another
+  good alternative is the usage of control flow `&&/||`.
+<sup>[[link](#if-as-a-modifier)]</sup>
+
+  ```Ruby
+  # bad
+  if some_condition
+    do_something
+  end
+
+  # good
+  do_something if some_condition
+
+  # another good option
+  some_condition && do_something
+  ```
+
+* <a name="no-multiline-if-modifiers"></a>
+  Avoid modifier `if/unless` usage at the end of a non-trivial multi-line
+  block.
+<sup>[[link](#no-multiline-if-modifiers)]</sup>
+
+  ```Ruby
+  # bad
+  10.times do
+    # multi-line body omitted
+  end if some_condition
+
+  # good
+  if some_condition
+    10.times do
+      # multi-line body omitted
+    end
+  end
+  ```
+
+* <a name="unless-for-negatives"></a>
+
+Avoid usage of `unless` over negative conditions (or control flow), it
+can cause issues when having multi term expressions.
+<sup>[[link](#unless-for-negatives)]</sup>
+
+  ```Ruby
+  # bad
+  do_something unless some_condition
+
+  # good
+  do_something if !some_condition
+
+  # another good option
+  some_condition || do_something
+  ```
+
+* <a name="no-else-with-unless"></a>
+  Do not use `unless` with `else`. Rewrite these with the positive case first.
+<sup>[[link](#no-else-with-unless)]</sup>
+
+  ```Ruby
+  # bad
+  unless success?
+    puts 'failure'
+  else
+    puts 'success'
+  end
+
+  # good
+  if success?
+    puts 'success'
+  else
+    puts 'failure'
+  end
+  ```
+
+* <a name="no-parens-if"></a>
+  Don't use parentheses around the condition of an `if/unless/while/until`.
+<sup>[[link](#no-parens-if)]</sup>
+
+  ```Ruby
+  # bad
+  if (x > 10)
+    # body omitted
+  end
+
+  # good
+  if x > 10
+    # body omitted
+  end
+  ```
+
+Note that there is an exception to this rule, namely [safe assignment in
+condition](#safe-assignment-in-condition).
+
+* <a name="no-multiline-while-do"></a>
+  Do not use `while/until condition do` for multi-line `while/until`.
+<sup>[[link](#no-multiline-while-do)]</sup>
+
+  ```Ruby
+  # bad
+  while x > 5 do
+    # body omitted
+  end
+
+  until x > 5 do
+    # body omitted
+  end
+
+  # good
+  while x > 5
+    # body omitted
+  end
+
+  until x > 5
+    # body omitted
+  end
+```
+
+* <a name="infinite-loop"></a>
+  Use `Kernel#loop` instead of `while/until` when you need an infinite loop.
```

Comment:
Style-wise, I agree. In practice, however, `while true` is faster. An experiment I did last year - https://github.com/jordansissel/experiments/blob/master/ruby/jruby-loop/test.rb

`loop do ... end` sets up a block and a new context, and I think that's why it's slower, but that's just guessing. Proof:

* Jruby 1.7.19

```
% ruby test.rb
Rehearsal ------------------------------------------------------------------
loop do                          5.670000   0.260000   5.930000 (  6.030000)
while true                       1.860000   0.050000   1.910000 (  1.895000)
--------------------------------------------------------- total: 7.840000sec

                                     user     system      total        real
loop do                          5.000000   0.110000   5.110000 (  5.591000)
while true                       2.220000   0.060000   2.280000 (  2.658000)
```

* MRI 2.1.6

```
ruby test.rb
Rehearsal ------------------------------------------------------------------
loop do                          7.680000   0.100000   7.780000 (  8.075701)
while true                       2.440000   0.030000   2.470000 (  2.642652)
-------------------------------------------------------- total: 10.250000sec

                                     user     system      total        real
loop do                          7.500000   0.070000   7.570000 (  7.678882)
while true                       2.420000   0.010000   2.430000 (  2.443713)
```